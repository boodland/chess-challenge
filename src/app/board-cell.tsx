import { memo } from 'react';
import styles from './board-cell.module.css';
import { CellTypeEnum, getShape } from './utils';

export interface BoardCellProps {
  piece: string;
  type: CellTypeEnum;
  blocked?: boolean;
  selected?: boolean;
  onClick: () => void;
}

const BoardCellMemoized = memo(
  ({ piece, type, blocked, selected, onClick }: BoardCellProps) => {
    const blockedClass = blocked ? styles[`blocked`] : '';
    const selectedClass = selected ? styles[`selected`] : '';
    const pieceShape = getShape(piece);

    return (
      <button
        className={`${styles['cell']} ${styles[type]} ${blockedClass} ${selectedClass}`}
        onClick={onClick}
      >
        {pieceShape}
      </button>
    );
  }
);

export default BoardCellMemoized;
