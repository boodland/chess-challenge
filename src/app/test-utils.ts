const PIECES = 'KQRBNP';

export const PieceShape = {
    'K': '♔',
    'Q': '♕',
    'R': '♖',
    'B': '♗',
    'N': '♘',
    'P': '♙',
    'k': '♚',
    'q': '♛',
    'r': '♜',
    'b': '♝',
    'n': '♞',
    'p': '♟',
    '': '',
};

export const EMPTY_PIECE = PieceShape[''];

const getRandomPiece = () => PIECES.charAt(Math.floor(Math.random() * PIECES.length));
export const getRandomWhitePiece = getRandomPiece;
export const getRandomBlackPiece = () => getRandomPiece().toLowerCase();

export const EMPTY_HANDLER = () => null;

