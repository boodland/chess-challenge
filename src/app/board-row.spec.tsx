import { render } from '@testing-library/react';
import { BoardCellProps } from './board-cell';
import BoardRow from './board-row';
import { INITIAL_POSITIONS } from './test-utils';

const nullFunction = () => null;

vi.mock('./utils', async (importActual) => {
  const actual = (await importActual()) as typeof import('./utils');
  return {
    ...actual,
    isEmpty: () => false,
    COLUMNS: ['a', 'b', 'c'],
  };
});

vi.mock('react', async (importActual) => {
  const actual = (await importActual()) as typeof import('react');
  return {
    ...actual,
    useContext: () => ({
      positions: INITIAL_POSITIONS,
      selectedPosition: undefined,
      onClickPosition: nullFunction,
    }),
  };
});

vi.mock('./board-cell', () => ({
  default: (props: BoardCellProps) =>
    `MockedBoardCell ${JSON.stringify(props)} onClick: ${props.onClick}`,
}));

describe('BoardRow', () => {
  it('should render successfully', () => {
    const { baseElement } = render(<BoardRow startPosition={0} />);
    expect(baseElement).toBeTruthy();
  });

  it('should display one BoardCell per column', () => {
    const { container } = render(<BoardRow startPosition={0} />);
    expect(container).toMatchSnapshot();
  });

  describe('should pass the correct params to BoardCells', () => {
    it('when starting a different row', () => {
      const { container } = render(<BoardRow startPosition={1} />);
      expect(container).toMatchSnapshot();
    });
    it('when selectedPosition is set', () => {
      vi.mock('react', async (importActual) => {
        const actual = (await importActual()) as typeof import('react');
        return {
          ...actual,
          useContext: () => ({
            positions: INITIAL_POSITIONS,
            selectedPosition: 0,
            onClickPosition: nullFunction,
          }),
        };
      });
      const { container } = render(<BoardRow startPosition={0} />);
      expect(container).toMatchSnapshot();
    });
    it('when positions are empty and no selectedPosition', () => {
      vi.mock('./utils', async (importActual) => {
        const actual = (await importActual()) as typeof import('./utils');
        return {
          ...actual,
          isEmpty: () => true,
          COLUMNS: ['a', 'b', 'c'],
        };
      });
      vi.mock('react', async (importActual) => {
        const actual = (await importActual()) as typeof import('react');
        return {
          ...actual,
          useContext: () => ({
            positions: INITIAL_POSITIONS,
            selectedPosition: undefined,
            onClickPosition: nullFunction,
          }),
        };
      });
      const { container } = render(<BoardRow startPosition={0} />);
      expect(container).toMatchSnapshot();
    });
  });
});
