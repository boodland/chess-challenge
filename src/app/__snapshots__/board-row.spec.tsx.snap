// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BoardRow > should display one BoardCell per column 1`] = `
<div>
  <div>
    MockedBoardCell {"piece":"r","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"n","type":"light","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"b","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
  </div>
</div>
`;

exports[`BoardRow > should pass the correct params to BoardCells > when positions are empty and no selectedPosition 1`] = `
<div>
  <div>
    MockedBoardCell {"piece":"r","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"n","type":"light","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"b","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
  </div>
</div>
`;

exports[`BoardRow > should pass the correct params to BoardCells > when selectedPosition is set 1`] = `
<div>
  <div>
    MockedBoardCell {"piece":"r","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"n","type":"light","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"b","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
  </div>
</div>
`;

exports[`BoardRow > should pass the correct params to BoardCells > when starting a different row 1`] = `
<div>
  <div>
    MockedBoardCell {"piece":"q","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"k","type":"light","blocked":true,"selected":false} onClick: () =&gt; void 0
    MockedBoardCell {"piece":"b","type":"dark","blocked":true,"selected":false} onClick: () =&gt; void 0
  </div>
</div>
`;
